cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/AuthenticationContextValidator.java

```diff
@@ -0,0 +1,191 @@
+            final Optional<MultifactorAuthenticationProvider> result = Arrays.stream(providers)
+                    .filter(provider -> provider.equals(requestedProvider) || provider.getOrder() >= requestedProvider.get().getOrder())
+                    .findFirst();
+
+            if (result.isPresent()) {
+                logger.debug("Current provider {} already satisfies the authentication requirements of {}; proceed with flow normally.",
+                        result.get(), requestedProvider);
+                return new Pair(true, requestedProvider);
+            }
+        }
+
+        logger.debug("No multifactor providers could be located to satisfy the requested context for {}", requestedProvider);
+
+        final RegisteredServiceMultifactorPolicy.FailureModes mode = getMultifactorFailureModeForService(service);
+        switch (mode) {
+            case PHANTOM:
+                if (!requestedProvider.get().verify(service)) {
+                    logger.debug("Service {} is configured to use a {} failure mode for multifactor authentication policy and "
+                                 + "since provider {} is unavailable at the moment, CAS will knowingly allow [{}] as a satisfied criteria "
+                                 + "of the present authentication context",
+                            mode, service.getServiceId(), requestedProvider);
+                    return new Pair(true, requestedProvider);
+                }
+                break;
+            case OPEN:
+                if (!requestedProvider.get().verify(service)) {
+                    logger.debug("Service {} is configured to use a {} failure mode for multifactor authentication policy and "
+                                 + "since provider {} is unavailable at the moment, CAS will consider the authentication satisfied "
+                                 + "without the presence of {}",
+                            mode, service.getServiceId(), requestedProvider);
+                    return new Pair(true, satisfiedProviders.stream().findFirst());
+                }
+                break;
+        }
+
+        return new Pair(false, requestedProvider);
+    }
+
+    /**
+     * Gets all multifactor authentication providers from application context.
+     *
+     * @return the all multifactor authentication providers from application context
+     */
+    private Map<String, MultifactorAuthenticationProvider> getAllMultifactorAuthenticationProvidersFromApplicationContext() {
+        try {
+            return this.applicationContext.getBeansOfType(MultifactorAuthenticationProvider.class);
+        } catch (final Exception e) {
+            logger.warn("Could not locate beans of type {} in the application context", MultifactorAuthenticationProvider.class);
+        }
+        return null;
+    }
+
+    private Collection<MultifactorAuthenticationProvider> getSatisfiedAuthenticationProviders(final Authentication authentication,
```

Comment:
Usually it _does_ improve the readability of code. Intention revealing,
declarative style of programming. It's of course understandable that for
any of us doing imperative style of Java for decades (at least couple of
decades for Java) :-) it's not clear right away how to solve problems with
this functional style (I'm including myself). It'll come, as we start doing
it over and over again. OK, nothing to see here, move alone :-)

On Sun, Feb 21, 2016 at 15:33 Daniel Frett <notifications@github.com> wrote:

> In
> cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/AuthenticationContextValidator.java
> <https://github.com/Jasig/cas/pull/1496#discussion_r53574991>:
>
> > +
> > +    /**
> > +     * Gets all multifactor authentication providers from application context.
> > +     *
> > +     * @return the all multifactor authentication providers from application context
> > +     */
> > +    private Map<String, MultifactorAuthenticationProvider> getAllMultifactorAuthenticationProvidersFromApplicationContext() {
> > +        try {
> > +            return this.applicationContext.getBeansOfType(MultifactorAuthenticationProvider.class);
> > +        } catch (final Exception e) {
> > +            logger.warn("Could not locate beans of type {} in the application context", MultifactorAuthenticationProvider.class);
> > +        }
> > +        return null;
> > +    }
> > +
> > +    private Collection<MultifactorAuthenticationProvider> getSatisfiedAuthenticationProviders(final Authentication authentication,
>
> I'm +1 for streams when it doesn't leave the code less
> readable/maintainable than without streams ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Jasig/cas/pull/1496/files#r53574991>.
>